src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs

```diff
@@ -2018,5 +1995,58 @@ public void TestAccessibilityForPublicFields6()
 @"class A { public B<int> b = new B<int>(); public class B<T>{ public B(){}}}",
 index: 2);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostType()
+        {
+            Test(
+@"class Program { static void Main ( string [ ] args ) { [|MyCompany|] . MyProduct . MyArea . MyClass a ; } } ",
+@"namespace MyCompany.MyProduct.MyArea { internal class MyClass { } }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostTypeInsideNameOf()
+        {
+            Test(
+@"class Program { static void Main ( string [ ] args ) { var a = nameof ( [|MyCompany|] . MyProduct . MyArea . MyClass ) ; } } ",
+@"namespace MyCompany.MyProduct.MyArea { internal class MyClass { } }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostTypeNotPresentInConditional()
+        {
+            TestMissing(
+@"class Program { static void Main ( string [ ] args ) { [|MyCompany . MyProduct ? . MyArea . MyClass a |]; } }  ");
+        }
+
```

Comment:
Yeah, I should read first before posting...

I think we should offer to generate a type for `MyProduct` in namespace `MyCompany`